packages/test/local-server-tests/src/test/handleResolution.spec.ts

```diff
@@ -0,0 +1,365 @@
+			throw responseToException(response, request);
+		}
+		return response.value as ITestFluidObject;
+	} catch (e) {
+		return Promise.reject(e);
+	}
+}
+
+/**
+ * Creates a non-root data object and validates that it is not visible from the root of the container.
+ */
+async function createNonRootDataObject(
+	containerRuntime: IContainerRuntimeBase,
+): Promise<ITestFluidObject> {
+	const dataStore = await containerRuntime.createDataStore("default");
+	const dataObject = await getDataStoreEntryPointBackCompat<ITestFluidObject>(dataStore);
+	// Non-root data stores are not visible (unreachable) from the root unless their handles are stored in a
+	// visible DDS.
+	await assert.rejects(
+		resolveHandleWithoutWait(containerRuntime, dataObject.context.id),
+		"Non root data object must not be visible from root after creation",
+	);
+	return dataObject;
+}
+
+async function getAndValidateDataObject(
+	fromDataObject: ITestFluidObject,
+	key: string,
+): Promise<ITestFluidObject> {
+	const dataObjectHandle = fromDataObject.root.get<IFluidHandle<ITestFluidObject>>(key);
+	assert(dataObjectHandle !== undefined, `Data object handle for key ${key} not found`);
+	const dataObject = await dataObjectHandle.get();
+	const runtime = dataObject.context.containerRuntime;
+	await assert.doesNotReject(
+		resolveHandleWithoutWait(runtime, dataObject.context.id),
+		`Data object for key ${key} must be visible`,
+	);
+	return dataObject;
+}
+
+/**
+ * Validates that non-root data stores that have other non-root data stores as dependencies are not visible
+ * until the parent data store is visible. Also, they are visible in remote clients and can send ops.
+ */
+describe("multi-level object visibility tests", () => {
+	const documentId = "objectVisibilityTest";
+	const documentLoadUrl = `https://localhost/${documentId}`;
+	let container1: IContainer;
+	let containerRuntime1: IContainerRuntimeBase;
+	let dataObject1: ITestFluidObject;
+	let documentServiceFactory: LocalDocumentServiceFactory;
+	let deltaConnectionServer: ILocalDeltaConnectionServer;
+	const urlResolver = new LocalResolver();
+
+	const defaultFactory: TestFluidObjectFactory = new TestFluidObjectFactory(
+		[[mapId, SharedMap.getFactory()]],
+		"default",
+	);
+	const runtimeFactory = new ContainerRuntimeFactoryWithDefaultDataStore({
+		defaultFactory,
+		registryEntries: [[defaultFactory.type, Promise.resolve(defaultFactory)]],
+	});
+
+	const loaderContainerTracker = new LoaderContainerTracker();
+
+	function getDetachedContainerProps(): ICreateDetachedContainerProps {
+		const codeDetails: IFluidCodeDetails = {
+			package: "test",
+			config: {},
+		};
+		const codeLoader = new LocalCodeLoader([[codeDetails, runtimeFactory]]);
+
+		deltaConnectionServer = LocalDeltaConnectionServer.create();
+
+		documentServiceFactory = new LocalDocumentServiceFactory(deltaConnectionServer);
+
+		return {
+			urlResolver,
+			documentServiceFactory,
+			codeLoader,
+			codeDetails,
+		};
+	}
```

Comment:
there is an existing helper function that will handle most of this, and get ride of even more shared state. just start tests with something like 
```
		const loaderContainerTracker = new LoaderContainerTracker();
		const deltaConnectionServer = LocalDeltaConnectionServer.create();

		const loader = createLoader({
			deltaConnectionServer,
		});
		const container1 = await createDetachedContainer(loader);
```